src/box/memtx_engine.cc

```diff
@@ -2032,18 +2033,20 @@ memtx_tuple_new_raw_impl(struct tuple_format *format, const char *data,
 		goto end;
 	}
 
-	while ((tuple = MemtxAllocator<ALLOC>::alloc_tuple(total)) == NULL) {
+	while ((block = MemtxAllocator<ALLOC>::alloc(tuple_len, data_offset,
+						     make_compact)) == NULL) {
 		bool stop;
 		memtx_engine_run_gc(memtx, &stop);
 		if (stop)
 			break;
 	}
-	if (tuple == NULL) {
+	if (block == NULL) {
 		diag_set(OutOfMemory, total, "slab allocator", "memtx_tuple");
 		goto end;
 	}
-	tuple_create(tuple, 0, tuple_format_id(format),
-		     data_offset, tuple_len, make_compact);
+
+	tuple = memtx_block_to_tuple(block);
+	tuple_create_base(tuple, 0, tuple_format_id(format));
```

Comment:
Nit: we should describe this contract somewhere in `memtx_allocator`. I mean, that allocated tuple has `data_offset` and `size` fields set, so one should call `tuple_create_base` instead of `tuple_create` to initialize tuple completely. As for me, description of `memtx_block_to_tuple` or `memtx_block::header` would be perfect.

Reply:
Added comment to `memtx_block::header`.